[nnc] Don't use sleef where it's slower

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51252 [nnc] Don't fuse intrinsics on cpu if they are slower
* **#51246 [nnc] Don't use sleef where it's slower**
* #51190 [nnc] Expose vectorized math functions to jit fuser.
* #51187 [nnc] Use sleef if its symbols are available
* #51186 [nnc][trivial] Refactor llvm_jit so the wrapper class doesn't depend on ifdefs
* #51125 [nnc] Refactor generation of intrinsics to reduce the amount of macro-hell

Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

bertmaher committed:
[nnc] Don't use sleef where it's slower

Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
🚢

bertmaher committed:
Fix a super persnickity build warning/error about throw specifiers on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

bertmaher committed:
Really fix persnickity build warning on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

bertmaher committed:
lint on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

bertmaher committed:
ifdef x86_64 only on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

bertmaher committed:
Work around debug-mode llvm really not wanting you to lookup undefined symbols on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

bertmaher committed:
Forgot to insert the symbols in the set... on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -1321,6 +1321,23 @@ void LLVMCodeGenImpl::emitIsNan(const Intrinsics* v) {
+      "fabsf",
+      "floorf",
+  };
+  return noSleef.find(name) == noSleef.end();
```
nit: i find `noSleef.count(name) == 0` to be a little easier to parse personally (not worth updating this PR just for this comment ofc)

bertmaher replied:
Haha, I do too generally, I have no idea what possessed me to do `find == end`.

bertmaher committed:
fix sleef type suffix at the appropriate point in the stack on "[nnc] Don't use sleef where it's slower"


Using sleef is sometimes slower than libm (I haven't debugged why).
The easy solution is to not use sleef in those cases.  With this diff, plus the
prior one to use sleef period, we've sped up every unary op:

Differential Revision: [D26113672](https://our.internmc.facebook.com/intern/diff/D26113672/)

[ghstack-poisoned]
